internal/engine/wazevo/ssa/pass.go

```diff
@@ -348,6 +330,138 @@ func passNopInstElimination(b *builder) {
 						b.alias(cur.Return(), x)
 					}
 				}
+			// Z := Const 0
+			// (Iadd X, Z) => X
+			// (Iadd Z, Y) => Y
+			case OpcodeIadd:
+				x, y := cur.Arg2()
+				definingInst := b.valueIDToInstruction[y.ID()]
+				if definingInst == nil {
+					if definingInst = b.valueIDToInstruction[x.ID()]; definingInst == nil {
+						continue
+					} else {
+						x = y
+					}
+				}
+				if definingInst.Constant() && definingInst.ConstantVal() == 0 {
+					b.alias(cur.Return(), x)
+				}
+			}
+		}
+	}
+}
+
+func passCollectValueIdToInstructionMapping(b *builder) {
+	if int(b.nextValueID) >= len(b.valueIDToInstruction) {
+		b.valueIDToInstruction = append(b.valueIDToInstruction, make([]*Instruction, b.nextValueID)...)
+	}
+
+	for blk := b.blockIteratorBegin(); blk != nil; blk = b.blockIteratorNext() {
+		for cur := blk.rootInstr; cur != nil; cur = cur.next {
+			r1, rs := cur.Returns()
+			if r1.Valid() {
+				b.valueIDToInstruction[r1.ID()] = cur
+			}
+			for _, r := range rs {
+				b.valueIDToInstruction[r.ID()] = cur
+			}
+		}
+	}
+}
+
+// passConstFoldingOpt scans all instructions for arithmetic operations over constants,
+// and replaces them with a const of their result.
+func passConstFoldingOpt(b *builder) {
+	for blk := b.blockIteratorBegin(); blk != nil; blk = b.blockIteratorNext() {
+		for cur := blk.rootInstr; cur != nil; cur = cur.next {
+			// The fixed point is reached through a simple iteration over the list of instructions.
+			// Note: Instead of just an unbounded loop with a flag, we may also add an upper bound to the number of iterations.
+			isFixedPoint := false
+			for !isFixedPoint {
+				isFixedPoint = true
+				op := cur.Opcode()
+				switch op {
+				// X := Const xc
+				// Y := Const yc
+				// - (Iadd X, Y) => Const (xc + yc)
+				case OpcodeIadd, OpcodeIsub, OpcodeImul:
+					x, y := cur.Arg2()
+					xDef := b.valueIDToInstruction[x.ID()]
+					yDef := b.valueIDToInstruction[y.ID()]
+					if xDef == nil || yDef == nil {
+						// If we are adding some parameter, ignore.
+						continue
+					}
+					if xDef.Constant() && yDef.Constant() {
+						isFixedPoint = false
+						// Mutate the instruction to an Iconst.
+						cur.opcode = OpcodeIconst
+						// Clear the references to operands.
+						cur.v, cur.v2 = ValueInvalid, ValueInvalid
+						// We assume all the types are consistent.
+						if x.Type().Bits() == 64 {
+							xc, yc := int64(xDef.ConstantVal()), int64(yDef.ConstantVal())
+							switch op {
+							case OpcodeIadd:
+								cur.u1 = uint64(xc + yc)
+							case OpcodeIsub:
+								cur.u1 = uint64(xc - yc)
+							case OpcodeImul:
+								cur.u1 = uint64(xc * yc)
```

Comment:
I would like to see the overflow is handled correctly for 32-bits in a unit test